test/crowdsale/ZxcCrowdsale.test.js

```diff
@@ -580,328 +577,311 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
+                                           startTimeSaleNoBonus,
+                                           endTime,
+                                           rate,
+                                           crowdsaleCap,
+                                           crowdsaleCap,
+                                           bonusPresale,
+                                           bonusSale,
+                                           minimumPresaleWeiDeposit,
+                                           {from: crowdsaleOwner});
+  
+        await token.approve(crowdsale.address, crowdsaleCap, {from: tokenOwner});
+        await token.setCrowdsaleAddress(crowdsale.address, {from: tokenOwner});
+        await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+  
+        await assertRevert(crowdsale.buyTokens({from: buyer, value: weiAmount}));
+      });
+  
+      it('buyTokens should fail purchasing tokens if transferFrom fails', async () => {
+        const weiAmount = ether(2.1);
+  
+        await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+  
+        token.approve(crowdsale.address, 0, {from: tokenOwner});
+        await assertRevert(crowdsale.buyTokens({from: buyer, value: weiAmount}));
+      });
+
+      it('fallback function should purchase tokens', async () => {
+        const weiAmount = ether(8.05113);
+        const expectedSoldTokens = weiAmount.mul(rate);
+        const startWalletBalance = await web3.eth.getBalance(wallet);
+  
+        await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+  
+        const { logs } = await crowdsale.sendTransaction({from: buyer, value: weiAmount});
+        const actualBalance = await token.balanceOf(buyer);
+        // Buyer should get correct number of tokens
+        assert.equal(actualBalance.toString(), expectedSoldTokens.toString());
+        // Wallet should receive correct amount of wei
+        const endWalletBalance = await web3.eth.getBalance(wallet);
+        assert.strictEqual(endWalletBalance.sub(startWalletBalance).toString(), weiAmount.toString());
+        // Counter for sold ZXC should be increased
+        const zxcSold = await crowdsale.zxcSold.call()
+        assert.strictEqual(zxcSold.toString(), expectedSoldTokens.toString());
+  
+        const event = logs.find(e => e.event === 'TokenPurchase');
+        assert.notEqual(event, undefined);
+      });
+    });
+
+    describe('KYC level 1', function() {
+      beforeEach(async () => {
+        data = [web3Util.padLeft(web3Util.numberToHex(1), 64)];
+  
+        await xcertToken.mint(buyer,
+                              123,
+                              "https://foobar.io",
+                              "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
+                              config,
+                              data,
+                              {from: xcertTokenOwner});
+      });
+
+      it('buyTokens should fail in presale', async () => {
```

Comment:
should fail prior, presale, after. Should succeed in crowdsale /w bonus and w/o bonus.